src/services/datasource.ts

```diff
@@ -25,26 +44,82 @@ class WrappedLokiDatasource extends RuntimeDataSource<DataQuery> {
       getDataSourceSrv()
         .get(getDataSource(request.scopedVars.__sceneObject.valueOf()))
         .then((ds) => {
-          // override the target datasource to Loki
-          request.targets = request.targets.map((target) => {
-            target.datasource = ds;
-            return target;
-          });
-
-          // query the datasource and return either observable or promise
-          const dsResponse = ds.query(request);
-          if (isObservable(dsResponse)) {
-            dsResponse.subscribe(subscriber);
-          } else {
-            dsResponse.then((response) => {
-              subscriber.next(response);
-              subscriber.complete();
-            });
+          if (!(ds instanceof DataSourceWithBackend)) {
+            throw new Error('Invalid datasource!');
+          }
+
+          const requestType = request.targets?.[0]?.resource;
+
+          switch (requestType) {
+            case 'volume': {
+              this.transformVolumeResponse(request, ds, subscriber);
+              break;
+            }
+            case 'patterns': {
+              console.warn('not yet implemented');
+              // this.transformPatternResponse(request, ds, subscriber);
+              break;
+            }
+            default: {
+              // override the target datasource to Loki
+              request.targets = request.targets.map((target) => {
+                target.datasource = ds;
+                return target;
+              });
+
+              // query the datasource and return either observable or promise
+              const dsResponse = ds.query(request);
+              dsResponse.subscribe(subscriber);
+              break;
+            }
           }
         });
     });
   }
 
+  private transformVolumeResponse(
+    request: DataQueryRequest<LokiQuery & SceneDataQueryResourceRequest>,
+    ds: DataSourceWithBackend<LokiQuery>,
+    subscriber: Subscriber<DataQueryResponse>
+  ) {
+    if (request.targets.length !== 1) {
+      throw new Error('Volume query can only have a single target!');
+    }
+
+    const targetsInterpolated = ds.interpolateVariablesInQueries(request.targets, request.scopedVars);
+
+    const dsResponse = ds.getResource(
+      'index/volume',
+      {
+        expr: targetsInterpolated[0].expr,
+        query: targetsInterpolated[0].expr,
+        from: request.range.from.utc().toISOString(),
+        to: request.range.to.utc().toISOString(),
+        limit: 1000,
+      },
+      {
+        requestId: request.requestId ?? 'volume',
+        headers: {
+          'X-Query-Tags': `Source=${PLUGIN_ID}`,
+        },
+      }
+    );
+    dsResponse.then((response: IndexVolumeResponse | undefined) => {
+      response?.data.result.sort((lhs, rhs) => {
+        return Number(rhs.value[1]) - Number(lhs.value[1]);
+      });
```

Comment:
Can we get some comment that `value[0]` is the timestamp and `value[1]` is the value?

Reply:
we should have a type somewhere that describes this, I'll add a type assertion or something